### DISCUSSION TITLE
[RFC] Semantic DOM Structure for all Components

### DISCUSSION THREAD

**Discussion by MadCcc** [ant-design/ant-design#40221] [RFCs]
## Summary

Define semantic dom structure, .e.g `root` `inner`. These conceptions could be used in `classes` `styles` or other dom related props to specify which element it would be applied to.

## Activation

Currently we may provide multiple `-ClassName` or `-Style` props for one component, such as `popupClassName` `rootClassName`. These props are usually named with element type or usage and would become more in the future. So we prefer to merge theme into `classes` or `styles`. For example, we could use `popupClassName="custom-popup"` before, and now it should be conterted to `classes={{ popup: 'custom-popup' }}`, which is simpler and more explict.

In order to introduce `classes` `styles` and other props like this, we need to define semantic dom structure for each component. The structure may be different from each other, so each component in antd would have unique type defination for these props. And each property, such as `root`, would be correspond with one element in the component.

## API

**Tooltip**

| Property | Description | Type | Default | 
| --- | --- | --- | --- | 
| classes | Additional classes in Select | `{ root?: string; inner?: string }` | - | 
| styles | Additional styles in Select | `{ root?: string; inner?: string }` | - | 

## Basic Example

```tsx
import React from 'react';
import { Tooltip } from 'antd';

export default () => (
  <Tooltip title="Example text." classes={{ root: 'custom-root', inner: 'custom-inner' }}>
    <button>Button</button>
  </Tooltip>
)
```

## Detailed Design

Since components have different dom structure or implementation, we would introduce `classes` and `styles` one by one. If there are already some `-ClassName` and `-Styles` props, we should merge them into `classes` and `styles`.

Common naming of semantic dom structure would be listed here and in wiki, and would be updated at any time if needed.

| Property | Description | 
| --- | --- |
| root | The root element of component, and would be different in some case that dom would changed by props. Every component should have this property. |
| inner | Element with class `*-inner`. |
| popup | The popup element. Would be used in some components which have popups, .e.g Select. |

More semantic properties could be introduce in the future. Once one is added, all dom-related props should support it.

## How we teach this

In **feature** branch:

1. Create new table in document for semantic dom structure for each component.
2. Introduce `classes` and `styles` into components whose structure have been discussed and approved by collaborators.
3. Deprecate all `-ClassName` and `-Styles` and give warning.

**Comment by kiner-tang** [comment]
深有同感，以前各种class和style实在是太松散了，感觉聚合起来更加方便管理，使用起来也更加方便

**Comment by arvinxx** [comment]
stye  ->  styles
classes -> classNames

建议命名对其

**Comment by GarvitMehta44** [comment]
import React from 'react';
import { Tooltip } from 'antd';

export default () => (
  <Tooltip title="Example text." classes={{ root: 'custom-root', inner: 'custom-inner' }}>
    <button>Button</button>
  </Tooltip>
)

**Comment by arvinxx** [comment]
这里是不是应该做个汇总，看下哪些已经实现了，哪些没实现？

**Comment by thinkasany** [comment]
所以 tooltip 语义化最终的 api 是敲定 root 和 inner 了嘛
![image](https://github.com/user-attachments/assets/1b585ee6-f185-4e36-bec2-b8073942fef3)

我想在 https://github.com/ant-design/ant-design/pull/51814 中实现它，以在 ConfigProvider 中能够配置。

**Comment by coding-ice** [comment]
现在的 `classNames` 或者 `styles`, 这个顺序也应该统一一下，不然同等权重下覆盖起来就不统一了

比如：
```js
const { button } = useContext(ConfigContext);
const cls = classNames(button.className, className, rootClassName);
```
